xla/service/reduce_scatter_all_gather_combiner_test.cc

```diff
@@ -0,0 +1,184 @@
+
+namespace xla {
+namespace {
+
+namespace op = xla::testing::opcode_matchers;
+
+class ReduceScatterAllGatherCombinerTest : public HloTestBase {
+ public:
+  StatusOr<std::unique_ptr<HloModule>> RunPass(absl::string_view hlo_module,
+                                               int64_t num_replicas,
+                                               int64_t num_partitions,
+                                               bool expect_change) {
+    HloModuleConfig config = GetModuleConfigForTest(
+        /*replica_count=*/num_replicas,
+        /*num_partitions=*/num_partitions);
+    config.set_use_spmd_partitioning(num_partitions > 1);
+    TF_ASSIGN_OR_RETURN(auto module,
+                        ParseAndReturnVerifiedModule(hlo_module, config));
+
+    auto changed = ReduceScatterAllGatherCombiner().Run(module.get());
+    if (!changed.ok()) {
+      return changed.status();
+    }
+    EXPECT_EQ(changed.value(), expect_change);
+    return StatusOr<std::unique_ptr<HloModule>>(std::move(module));
+  }
+
+  template <HloOpcode op>
+  size_t CollectiveCount(std::unique_ptr<HloModule> &module) {
+    return absl::c_count_if(module->entry_computation()->instructions(),
+                            HloPredicateIsOp<op>);
+  }
+};
+
+TEST_F(ReduceScatterAllGatherCombinerTest, SimpleCombine) {
+  absl::string_view hlo_string = R"(
+HloModule ReduceScatter
+
+add {
+  x = bf16[] parameter(0)
+  y = bf16[] parameter(1)
+  ROOT add = bf16[] add(x, y)
+}
+
+ENTRY main {
+param.1 = bf16[8,128,1024] parameter(0)
+reduce-scatter.1 = bf16[8,64,1024] reduce-scatter(param.1), channel_id=8, replica_groups={{0,1},{2,3},{4,5},{6,7}}, use_global_device_ids=true, dimensions={1}, to_apply=add
+all-gather.1 = bf16[8,128,1024] all-gather(reduce-scatter.1), channel_id=5, replica_groups={{0,1},{2,3},{4,5},{6,7}}, dimensions={1}, use_global_device_ids=true
+}
+)";
+
+  TF_ASSERT_OK_AND_ASSIGN(auto module, RunPass(hlo_string,
+                                               /*num_replicas=*/8,
+                                               /*num_partitions=*/1,
+                                               /*expect_change=*/true));
+  EXPECT_EQ(CollectiveCount<HloOpcode::kAllGather>(module), 0);
+  EXPECT_EQ(CollectiveCount<HloOpcode::kReduceScatter>(module), 0);
+  EXPECT_EQ(CollectiveCount<HloOpcode::kAllReduce>(module), 1);
+}
+
+TEST_F(ReduceScatterAllGatherCombinerTest, IncompatibleProperties) {
+  absl::string_view hlo_string = R"(
+HloModule ReduceScatter
+
+add {
+  x = bf16[] parameter(0)
+  y = bf16[] parameter(1)
+  ROOT add = bf16[] add(x, y)
+}
+
+ENTRY main {
+param.1 = bf16[8,128,1024] parameter(0)
+reduce-scatter.1 = bf16[8,64,1024] reduce-scatter(param.1), channel_id=8, replica_groups={{0,1},{2,3},{4,5},{6,7}}, use_global_device_ids=true, dimensions={1}, to_apply=add
+all-gather.1 = bf16[16,64,1024] all-gather(reduce-scatter.1), channel_id=5, replica_groups={{0,1},{2,3},{4,5},{6,7}}, dimensions={0}, use_global_device_ids=true
+}
```

Comment:
As mentioned above, please amend this test to test all incompatible properties as there are just a few. Here its dimensions, but we can extend it for other properties.

Reply:
added